.github/workflows/build.yaml

```diff
@@ -43,30 +78,150 @@ jobs:
+    timeout-minutes: 5
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v23.2
+        with:
+          files: |
+            **.sh
+            **.bats
+
+      - name: Install Node.js v16
+        if: steps.changed-files.outputs.any_changed == 'true'
+        uses: actions/setup-node@v3
+        with:
+          node-version: "16"
+
       - name: Fetch dependencies from cache
+        if: steps.changed-files.outputs.any_changed == 'true'
         id: cache-node-modules
         uses: actions/cache@v3
         with:
-          path: "**/node_modules"
-          key: yarn-build-${{ hashFiles('**/yarn.lock') }}
+          path: "node_modules"
+          key: yarn-build-${{ hashFiles('yarn.lock') }}
           restore-keys: |
             yarn-build-
 
       - name: Install dependencies
-        if: steps.cache-node-modules.outputs.cache-hit != 'true'
+        if: steps.changed-files.outputs.any_changed == 'true' && steps.cache-node-modules.outputs.cache-hit != 'true'
         run: SKIP_SUBMODULE_DEPS=1 yarn --frozen-lockfile
 
-      - name: Run yarn fmt
-        run: yarn fmt
-        if: success()
+      - name: Lint shell files
+        if: steps.changed-files.outputs.any_changed == 'true'
+        run: yarn lint:sh
 
-      - name: Run yarn lint
-        run: yarn lint
-        if: success()
+      - name: Fail workflow
+        if: failure()
+        uses: andymckay/cancel-action@0.2
+
+  lint-ts:
+    name: Lint TypeScript files
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v23.2
+        with:
+          files: |
+            **.ts
+            **.tsx
+            **.js
+          files_ignore: |
+            lib/vscode
+
+      - name: Install Node.js v16
+        if: steps.changed-files.outputs.any_changed == 'true'
+        uses: actions/setup-node@v3
+        with:
+          node-version: "16"
+
+      - name: Fetch dependencies from cache
+        if: steps.changed-files.outputs.any_changed == 'true'
+        id: cache-node-modules
+        uses: actions/cache@v3
+        with:
+          path: "node_modules"
+          key: yarn-build-${{ hashFiles('yarn.lock') }}
+          restore-keys: |
+            yarn-build-
+
+      - name: Install dependencies
+        if: steps.changed-files.outputs.any_changed == 'true' && steps.cache-node-modules.outputs.cache-hit != 'true'
+        run: SKIP_SUBMODULE_DEPS=1 yarn --frozen-lockfile
+
+      - name: Lint TypeScript files
+        if: steps.changed-files.outputs.any_changed == 'true'
+        run: yarn lint:ts
+
+      - name: Fail workflow
+        if: failure()
+        uses: andymckay/cancel-action@0.2
+
+  typecheck:
+    name: Check TypeScript types
```

Comment:
✍🏼 If we decide to move away from `ts-jest` we can modify this job to also typecheck `/test`